---
title: 'Lesson 10: Functions in R'
author: "Modesto"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Create user-written functions

## Why? How?

We have discussed thoroughout the last weeks how R can help you if you to save time when you need to analyze and plot the data from your experiment. However, many times, particularly in Bioinformatics, you won't have the data from one single experiment but from many of them.

Creating you own function will be very useful for automation of repetitive analyses or to encapsulate a sequence of expressions that need to be executed numerous times, perhaps under slightly different conditions. Functions are also often written when code must be shared with others or the public.

In R, functions are also considered as objects. That means that (1) they can be nested, so you can define a function inside another function and (2) you can use funcions as arguments to other functions. We will see very useful examples of this latter feature in Lesson 12, using custom functions as arguments for `lapply()` or `sapply()`.

The overall scheme of an R function is the following:

    my_function <- function(argument1, argument2,...){
      statements
      return(object)
      }

## My first function

Let's make a short function that convert R into a molecular biology dogma interpreter. It takes as input a nucleic acid sequence codon and returns its encoded amino acid in IUPAC one letter code.

```{r}
#my first function: the molecular biology dogma with R

codon2aa <- function(codon) {
  aa<-c()
  code<-read.csv2("data/genetic_code.csv")
  aa <- code$AA[code$Codon==codon]
  return(aa)
}

#now let's try it
codon2aa("ATG")
codon2aa("TAA")
codon2aa("CAT")
codon2aa("AXG")

```




```{r}

codon_to_aa <- function(codon) {
  aa<-c()
  code<-read.csv2("data/genetic_code.csv")
  aa <- code$AA[code$Codon==codon]
  if (length(aa) > 0){
    return(aa)
  }else{
    stop("Uno o más de los codones no es correcto. No se ha podido traducir.")  
  }
}

codon_to_aa("ATG")
aa #check aa, what's going on here?


#can we do it for more than one codon?? just loop it
codon_to_aa <- function(entrada) {
  aa<-c()
  code<-read.csv2("data/genetic_code.csv", stringsAsFactors = FALSE)
  for (i in 1:length(entrada)){
    if(entrada[i] %in% code$Codon){
      aa[i] <- code$AA[code$Codon==entrada[i]]
    } else {
      stop("Uno o más de los codones no es correcto. No se ha podido traducir.")
    }
  }
  return(aa)
}

codon_to_aa <- function(codon = A|C|G|T) {
  aa<-c()
  for (i in 1:length(codon)){
    code<-read.csv2("data/genetic_code.csv")
    aa[i] <- code$AA[code$Codon==codon[i]]
    if (length(aa) > 0{
      return(aa)
    }else{
      stop("Uno o más de los codones no es correcto. No se ha podido traducir.")  
    }
  }
  return(aa)
  }
  

codon_to_aa(c("ATG","TGA"))
codon_to_aa(c("ARG","TGA"))
codon_to_aa(c("ATG","CAT", "CAT","AAA","TAA"))

#now with a sequence
codon_to_aa <- function(orf) {

  if (nchar(orf) %% 3 != 0) {
    warning("La secuencia no es múltiplo de 3. Los últimos 1-2 nt no serán usados.")
  }
  codon <- c()
  for (i in 1:round(nchar(orf)/3)){
    codon[i]<-substr(orf,i,i+2)
    i=i+3
  }
  aa<-c()
  code<-read.csv2("data/genetic_code.csv")
  for (k in 1:length(codon)){
    if(codon[k] %in% code$Codon){
      aa[k] <- code$AA[code$Codon==codon[k]]
    } else {
      stop("Uno o más de los codones no es correcto. No se ha podido traducir.")
    }
  }
  paste(aa, collapse='')
}

codon_to_aa("ATGTGAAAA")
codon_to_aa("ATGTCGAAAA")
codon_to_aa("ATGTRGAAAA")



```

# References

-   *R programming for data science* <https://bookdown.org/rdpeng/rprogdatascience/>

-   Creating functions in *Programming in R* Swcarpentry <http://swcarpentry.github.io/r-novice-inflammation/02-func-R/index.html>

-   
